### PR TITLE
expression: compatibility is changed

### LINKED ISSUE #19150: Incompatibility with MySQL for ADDTIME()
### 1. Minimal reproduce step (Required)

```sql
select ADDTIME('2020-05-13 14:01:24', '2020-04-29 05:11:19');
```

### 2. What did you expect to see? (Required)
In MySQL 8.0.21

```sql
mysql> select ADDTIME('2020-05-13 14:01:24', '2020-04-29 05:11:19');
+-------------------------------------------------------+
| ADDTIME('2020-05-13 14:01:24', '2020-04-29 05:11:19') |
+-------------------------------------------------------+
| NULL                                                  |
+-------------------------------------------------------+
1 row in set (0.00 sec)
```

### 3. What did you see instead (Required)

```sql
mysql> select ADDTIME('2020-05-13 14:01:24', '2020-04-29 05:11:19');
+-------------------------------------------------------+
| ADDTIME('2020-05-13 14:01:24', '2020-04-29 05:11:19') |
+-------------------------------------------------------+
| 2020-05-13 19:12:43                                   |
+-------------------------------------------------------+
1 row in set (0.00 sec)
```

### 4. What is your TiDB version? (Required)

Release Version: v4.0.0-beta.2-944-gae779e150
Edition: Community
Git Commit Hash: ae779e1502cf99d65511387a4e1eb171bd9eb9c7
Git Branch: master
UTC Build Time: 2020-08-11 03:55:14
GoVersion: go1.13
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false

## SIG slack channel
[#sig-exec](https://slack.tidb.io/invite?team=tidb-community&channel=sig-exec&ref=high-performance)

## Score
- 300

## Mentor
 * @SunRunAway

### PR DESCRIPTION
**Pull request by Tangruilin**
### What problem does this PR solve?

Issue Number: close #19150 

Problem Summary:

### What is changed and how it works?
After `ParseDuration` in `builtinAddStringAndStringSig.evalString` , format like `*-*` will be checked，if the arg1Strmatch this format, it will return isNull

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [x] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

**Comment by Tangruilin** [comment]
when I pull the code, there are 4 test in expression cannot pass

**Review by zimulala — COMMENTED** [review]
```
[2021-12-12T13:04:46.677Z] --- FAIL: TestTimeBuiltin (0.51s)
[2021-12-12T13:04:46.677Z]     result.go:50:
[2021-12-12T13:04:46.677Z]         	Error Trace:	result.go:50
 	            				integration_serial_test.go:2170
[2021-12-12T13:04:46.677Z]         	Error:      	Not equal:
[2021-12-12T13:04:46.677Z]         	            	expected: "[2017-01-01 01:01:06 2017-01-01 01:00:56 2017-01-01 01:01:01 2017-01-01 01:01:02.340000]\n"
[2021-12-12T13:04:46.677Z]         	            	actual  : "[2017-01-01 01:01:06 <nil> 2017-01-01 01:01:01 2017-01-01 01:01:02.340000]\n"
```

**Review by dveeden — APPROVED** [review]
This seems to fix the issue described.

Note that there are still some other corner cases:

MySQL 8.0.27:
```
mysql> select ADDTIME('2020-05-13 14:01:24', '-10000000');
+---------------------------------------------+
| ADDTIME('2020-05-13 14:01:24', '-10000000') |
+---------------------------------------------+
| 2020-04-08 15:01:25                         |
+---------------------------------------------+
1 row in set, 1 warning (0.00 sec)

Warning (Code 1292): Truncated incorrect time value: '-10000000'
```

TiDB:
```
sql> select ADDTIME('2020-05-13 14:01:24', '-10000000');
+---------------------------------------------+
| ADDTIME('2020-05-13 14:01:24', '-10000000') |
+---------------------------------------------+
| NULL                                        |
+---------------------------------------------+
1 row in set, 1 warning (0.0007 sec)
Warning (code 1292): Truncated incorrect time value: '-10000000'
```

**Comment by lance6716** [comment]
ping @zimulala @bb7133

File: expression/builtin_time.go [hunk 1]
```diff
@@ -5556,6 +5557,16 @@ func (b *builtinAddStringAndStringSig) evalString(row chunk.Row) (result string,
 		}
 		return "", true, err
 	}
+
+	check := arg1Str
```
Comment (dveeden) [inline expression/builtin_time.go:5561]: I think it would be useful to add a comment here about why this is needed

Comment (Tangruilin) [reply expression/builtin_time.go:5561]: Yeah, In fact I think add the code in func `isDuration` or create a new func `maybeErrDuration` may be butter, but `isDuration` is a util func, I'm afraid to get new wrong, so I add the code here. I may add some cmment or create a new func, but which file may i add the func?

Comment (Tangruilin) [reply expression/builtin_time.go:5561]: I will check the other concor case and try to solve them

**Comment by Tangruilin** [comment]
ping @zimulala @bb7133

**Review by zimulala — APPROVED** [review]
LGTM

**Pull request merged**